akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -466,46 +464,53 @@ final class FoldAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends
             case Success(s) if s == null ⇒
               ReactiveStreamsCompliance.elementMustNotBeNullException
           }
+          val supervision = decider(ex)
 
-          if (decider(ex) == Supervision.Stop) failStage(ex)
-          else if (isClosed(in) && inFlight == 0) completeStage()
-          else if (!hasBeenPulled(in)) {
-            onResume(ex)
-            tryPull(in)
+          if (supervision == Supervision.Stop) {
+            aggregated.failure(ex)
+            failStage(ex)
+          } else {
+            if (supervision == Supervision.Restart) onRestart(ex)
+
+            aggregated.success(aggregator)
+
+            if (isAvailable(out) && !hasBeenPulled(in)) tryPull(in)
           }
         }
       }
     }).invoke _
 
-    def onPush(): Unit = {
+    def onPush(): Unit = if (aggregated.isCompleted) {
```

Comment:
We will only get a push if we did a pull, and we will only do a pull if the previous fold function has completed so this check is not needed.